clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -1146,6 +1160,9 @@ private PartitionRecords parseCompletedFetch(CompletedFetch completedFetch) {
                 log.warn("Received unknown topic or partition error in fetch for partition {}", tp);
                 this.metadata.requestUpdate();
             } else if (error == Errors.OFFSET_OUT_OF_RANGE) {
+                // Go back to the leader
+                subscriptions.clearPreferredReadReplica(tp);
```

Comment:
@hachikuji copying your comment from the other PR here:

```
Should the client act on the out of range error if we are not fetching from the current leader? I am wondering if the logic should be something like this:

if (fetchingFromLeader)
 normalOutOfRangeValidation
else
 clearPreferredReadReplica
 validateCurrentOffset
```

Reply:
Possibly we do not even need to validate the offset. We can just resume fetching from the leader.

Reply:
I was thinking the same. If we get out of range error from a follower and switch back to the leader, the offset should be valid still as long as the epoch hasn't changed.